### PR TITLE
[x] re-introduce check for overlay features, update for v2 resolver

### PR DESCRIPTION
**Pull request by sunshowers**
The Cargo v2 resolver can avoid unifying features for dev dependencies, so the
check can be relaxed a little.

This reverts commit 9a484bde and makes a minor change to it: the callback for the `resolve_with_fn` method is now conditional on `!link.dev_only()`.

### CODE REVIEW

**Comment by metajack** [comment]
Can you summarize what changed other than the revert? It's not clear from the diff which parts are new.

**Comment by sunshowers** [comment]
> Can you summarize what changed other than the revert? It's not clear from the diff which parts are new.

Ah sorry, just updated the summary. The line is at `// We now use the v2 resolver, so dev-only links can be skipped.`

File: devtools/x/src/lint/guppy.rs [hunk 1]
```diff
@@ -308,3 +309,108 @@ impl ProjectLinter for DirectDepDups {
...
+            .feature_graph()
+            .query_packages(&package_query, filter);
+
+        let mut overlays: Vec<(Option<&str>, &str, Option<&str>)> = vec![];
+
+        feature_query.resolve_with_fn(|_, link| {
+            // We now use the v2 resolver, so dev-only links can be skipped.
+            if !link.dev_only() {
+                // Consider the dependency even if it's dev-only since the v1 resolver unifies these.
+                // TODO: might be able to relax this for the v2 resolver.
```
Comment (metajack) [inline devtools/x/src/lint/guppy.rs:377]: This should probably get removed now :)

**Comment by metajack** [comment]
Did someone not check in a serde-generate update? Why is that showing up here.

**Comment by metajack** [comment]
Hmm, i wonder if it is because of the rebase. I'll have to check what the detected merge base sha is after bors pushes and see if I'm calculating that wrong.
